tools/unit/src/main/kotlin/com/datadog/tools/unit/ConditionWatcher.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.tools.unit
+
+import java.lang.RuntimeException
+
+/**
+ * Utility class which allows watch for the condition in the polling manner instead
+ * of using [Thread.sleep].
+ */
+class ConditionWatcher(
+    /**
+     * Polling interval in milliseconds, defaults to [DEFAULT_INTERVAL_MS].
+     */
+    private val pollingIntervalMs: Long = DEFAULT_INTERVAL_MS,
+    /**
+     * If it throws any sub-class of [AssertionError], condition won't fail immediately,
+     * but will poll once again. With any other [Throwable] produced by condition it will fail
+     * immediately.
+     */
+    private val condition: () -> Boolean
+) {
+
+    /**
+     * Waits for the condition to be satisfied. Will throw [TimeoutException] if it is not
+     * satisfied in the given timeout period.
+     * @param timeoutMs Maximum timeout in milliseconds.
+     */
+    fun doWait(timeoutMs: Long = DEFAULT_TIMEOUT_LIMIT_MS) {
+        var elapsedTime = 0L
+        var status = CONDITION_NOT_MET
+        var lastAssertionError: AssertionError? = null
+        do {
+            var invocationResult = false
+            try {
+                invocationResult = condition.invoke()
+            } catch (ae: AssertionError) {
+                lastAssertionError = ae
+            }
+
+            if (invocationResult) {
+                status = CONDITION_MET
+            } else {
+                elapsedTime += pollingIntervalMs
+                Thread.sleep(pollingIntervalMs)
+            }
+
+            if (elapsedTime >= timeoutMs) {
+                status = TIMEOUT
+                break
+            }
+        } while (status != CONDITION_MET)
+
+        if (status == TIMEOUT) {
+            reportTimeout(timeoutMs, lastAssertionError)
+        }
+    }
+
+    private fun reportTimeout(timeoutMs: Long, assertionError: AssertionError?) {
+        val message = "Waiting took more than ${timeoutMs / 1000} seconds. Test stopped."
+        if (assertionError == null) {
+            throw TimeoutException(message)
+        } else {
+            throw TimeoutException(
+                "$message Check failed assertion in the cause.",
```

Comment:
I think I will just redo this branch and simply throw `AssertionError` (without any timeout exception) which was recorded already, just to be nicer with IDE which applies special formatting to assertion errors.

Originally I added this because IDE collapses everything that it is not a part of the assertion error in the Run tab.

Reply:
Looks like when `AssertionError` is wrapped into another throwable, display in IDE is still good and we can see `expected + actual` message.